tiatoolbox/utils/image.py

```diff
@@ -20,17 +20,174 @@
 
 """Miscellaneous utilities which operate on image data."""
 import warnings
+from typing import Tuple, Union
 
 import numpy as np
 import cv2
 from PIL import Image
 
-from tiatoolbox.utils.transforms import bounds2locsize
-from tiatoolbox.utils.misc import conv_out_size, assert_dtype_int
+from tiatoolbox.utils.transforms import bounds2locsize, imresize
+from tiatoolbox.utils.misc import conv_out_size
+
+
+PADDING_TO_BOUNDS = np.array([-1, -1, 1, 1])
+"""
+Constant array which when multiplied with padding and added to bounds,
+applies the padding to the bounds.
+"""
+# Make this immutable / non-writable
+PADDING_TO_BOUNDS.flags.writeable = False
+
+
+def normalise_padding_size(padding):
+    """Normalises padding to be length 4 (left, top, right, bottom).
+
+    Given a scalar value, this is assumed to apply to all sides and
+    therefore repeated for each output (left, right, top, bottom). A
+    length 2 input is assumed to apply the same padding to the
+    left/right and top/bottom.
+
+    Args:
+        padding (int or tuple(int)): Padding to normalise.
+
+    Raises:
+        ValueError: Invalid input size of padding (e.g. length 3).
+        ValueError: Invalid input shape of padding (e.g. 3 dimensional).
+
+    Returns:
+        np.ndarray: Numpy array of length 4 with elements containing
+            padding for left, top, right, bottom.
+    """
+    padding_shape = np.shape(padding)
+    if len(padding_shape) > 1:
+        raise ValueError(
+            "Invalid input padding shape. Must be scalar or 1 dimentional."
+        )
+    padding_size = np.size(padding)
+    if padding_size == 3:
+        raise ValueError("Padding has invalid size 3. Valid sizes are 1, 2, or 4.")
+
+    if padding_size == 1:
+        padding = np.repeat(padding, 4)
+    elif padding_size == 2:
+        padding = np.tile(padding, 2)
+    else:
+        padding = np.array(padding)
+    return padding
+
+
+def make_bounds_size_positive(bounds):
+    """Make bounds have positive size and get horizontal/vertical flip flags.
+
+    Bounds with a negative size in either direction with have the
+    coordinates swapped (e.g. left and right or top and bottom swapped)
+    and a respective horizontal or vertical flip flag set in the
+    output to reflect the swaps which occurred.
+
+    Args:
+        bounds (np.ndarray): Length 4 array-like of bounds.
+
+    Returns:
+        tuple: Three tuple of: positive bounds, horizontal_flip boolean,
```

Comment:
`tuple(int, int, int)` ?

Reply:
This is a tough one. It looks to me as if there isn't a good way to document tuple returns using google docstring syntax. I can only find ways using restructured text or numpy style. One way I found its this:

```
    Returns:
        tuple: Three tuple containing positive bounds and flips:
            - tuple - Positive bounds
            - bool - Horizontal_flip
            - bool - Vertical_flip
```

which renders like this:
![image](https://user-images.githubusercontent.com/4615004/115538737-1a989980-a294-11eb-8677-955dad85a7a8.png)

You can also number them which is kind of neat:

![image](https://user-images.githubusercontent.com/4615004/115538881-3dc34900-a294-11eb-8974-3f81229381b4.png)

Reply:
This might be a nice way to do it:

```
    Returns:
        tuple: Three tuple containing positive bounds and flips:
            - :class:`numpy.ndarray` - Positive bounds
            - :py:obj:`bool` - Horizontal_flip
            - :py:obj:`bool` - Vertical_flip
```

![image](https://user-images.githubusercontent.com/4615004/115539683-29cc1700-a295-11eb-9b7c-7aeac6c4e745.png)

Reply:
better, let's use this from now on. may be a separate PR to clean this up.

Reply:
I also just noticed that this works but I think the above is more helpful / detailed:

![image](https://user-images.githubusercontent.com/4615004/115753780-772eae00-a393-11eb-9313-fe648c19259d.png)